File: ios-native/Dimo/Domain/DateHelpers.swift
import Foundation

enum RecurringOccurrenceSelection {
  case all, selected
}

enum DateHelpers {
  static let dayMS = 86_400_000

  static func localDateKey(_ date: Date, calendar: Calendar = .current) -> String {
    let c = calendar.dateComponents([.year, .month, .day], from: date)
    return String(format: "%04d-%02d-%02d", c.year ?? 0, c.month ?? 0, c.day ?? 0)
  }

  static func parseLocalDate(_ value: String, calendar: Calendar = .current) -> Date {
    let parts = value.split(separator: "-").compactMap { Int($0) }
    guard parts.count == 3 else { return Date() }
    var components = DateComponents()
    components.year = parts[0]
    components.month = parts[1]
    components.day = parts[2]
    return calendar.date(from: components) ?? Date()
  }

  static func formatTransactionDay(_ timestamp: Int, now: Date = Date(), calendar: Calendar = .current) -> String {
    let date = Date(timeIntervalSince1970: TimeInterval(timestamp) / 1000)
    let today = localDateKey(now, calendar: calendar)
    let key = localDateKey(date, calendar: calendar)
    if key == today { return "Today" }
    guard let yesterday = calendar.date(byAdding: .day, value: -1, to: calendar.startOfDay(for: now)) else {
      return key
    }
    if key == localDateKey(yesterday, calendar: calendar) { return "Yesterday" }

    let formatter = DateFormatter()
    formatter.locale = .current
    let sameYear = calendar.component(.year, from: date) == calendar.component(.year, from: now)
    formatter.setLocalizedDateFormatFromTemplate(sameYear ? "EEEE MMMd" : "EEEE MMMd yyyy")
    return formatter.string(from: date)
  }

  static func formatTransactionTime(_ timestamp: Int) -> String {
    let date = Date(timeIntervalSince1970: TimeInterval(timestamp) / 1000)
    let formatter = DateFormatter()
    formatter.locale = .current
    formatter.setLocalizedDateFormatFromTemplate("jmm")
    return formatter.string(from: date)
  }

  static func daysInMonth(year: Int, monthIndex: Int, calendar: Calendar = .current) -> Int {
    var components = DateComponents()
    components.year = year
    components.month = monthIndex + 1
    components.day = 1
    guard let start = calendar.date(from: components),
          let range = calendar.range(of: .day, in: .month, for: start) else { return 30 }
    return range.count
  }

  static func nextOccurrence(
    anchorDate: String,
    frequency: RecurringFrequency,
    now: Date = Date(),
    calendar: Calendar = .current
  ) -> Date {
    let anchor = parseLocalDate(anchorDate, calendar: calendar)
    let today = calendar.startOfDay(for: now)
    if anchor >= today { return anchor }

    let anchorDay = calendar.component(.day, from: anchor)
    let anchorMonth = calendar.component(.month, from: anchor) - 1

    if frequency == .monthly {
      let year = calendar.component(.year, from: today)
      let month = calendar.component(.month, from: today) - 1
      func candidate(year: Int, month: Int) -> Date {
        let day = min(anchorDay, daysInMonth(year: year, monthIndex: month, calendar: calendar))
        var c = DateComponents()
        c.year = year
        c.month = month + 1
        c.day = day
        return calendar.date(from: c) ?? today
      }
      var result = candidate(year: year, month: month)
      if result < today {
        var nextMonth = month + 1
        var nextYear = year
        if nextMonth > 11 {
          nextMonth = 0
          nextYear += 1
        }
        result = candidate(year: nextYear, month: nextMonth)
      }
      return result
    }

    func candidate(year: Int) -> Date {
      let day = min(anchorDay, daysInMonth(year: year, monthIndex: anchorMonth, calendar: calendar))
      var c = DateComponents()
      c.year = year
      c.month = anchorMonth + 1
      c.day = day
      return calendar.date(from: c) ?? today
    }
    let year = calendar.component(.year, from: today)
    var result = candidate(year: year)
    if result < today {
      result = candidate(year: year + 1)
    }
    return result
  }

  static func occurrencesThrough(
    anchorDate: String,
    frequency: RecurringFrequency,
    now: Date = Date(),
    calendar: Calendar = .current
  ) -> [Date] {
    let anchor = parseLocalDate(anchorDate, calendar: calendar)
    let today = calendar.startOfDay(for: now)
    if anchor > today { return [] }

    let day = calendar.component(.day, from: anchor)
    var dates: [Date] = []

    if frequency == .monthly {
      var year = calendar.component(.year, from: anchor)
      var month = calendar.component(.month, from: anchor) - 1
      while dates.count < 1200 {
        let clamped = min(day, daysInMonth(year: year, monthIndex: month, calendar: calendar))
        var c = DateComponents()
        c.year = year
        c.month = month + 1
        c.day = clamped
        guard let date = calendar.date(from: c) else { break }
        if date > today { break }
        dates.append(date)
        month += 1
        if month > 11 {
          month = 0
          year += 1
        }
      }
      return dates
    }

    var year = calendar.component(.year, from: anchor)
    let month = calendar.component(.month, from: anchor) - 1
    while dates.count < 200 {
      let clamped = min(day, daysInMonth(year: year, monthIndex: month, calendar: calendar))
      var c = DateComponents()
      c.year = year
      c.month = month + 1
      c.day = clamped
      guard let date = calendar.date(from: c) else { break }
      if date > today { break }
      dates.append(date)
      year += 1
    }
    return dates
  }

  static func recurringTransactionDates(
    anchorDate: String,
    frequency: RecurringFrequency,
    selection: RecurringOccurrenceSelection,
    now: Date = Date(),
    calendar: Calendar = .current
  ) -> [Date] {
    let anchor = parseLocalDate(anchorDate, calendar: calendar)
    let today = calendar.startOfDay(for: now)
    guard anchor <= today else { return [] }
    return selection == .all
      ? occurrencesThrough(anchorDate: anchorDate, frequency: frequency, now: now, calendar: calendar)
      : [anchor]
  }

  static func occurrenceTimestamp(_ date: Date, calendar: Calendar = .current) -> Int {
    var c = calendar.dateComponents([.year, .month, .day], from: date)
    c.hour = 12
    c.minute = 0
    c.second = 0
    let noon = calendar.date(from: c) ?? date
    return Int(noon.timeIntervalSince1970 * 1000)
  }

  static func occurrenceTimestamp(
    _ date: Date,
    time: Date,
    calendar: Calendar = .current
  ) -> Int {
    var dateParts = calendar.dateComponents([.year, .month, .day], from: date)
    let timeParts = calendar.dateComponents([.hour, .minute], from: time)
    dateParts.hour = timeParts.hour
    dateParts.minute = timeParts.minute
    dateParts.second = 0
    let combined = calendar.date(from: dateParts) ?? date
    return Int(combined.timeIntervalSince1970 * 1000)
  }

  static func recurringDueLabel(
    anchorDate: String,
    frequency: RecurringFrequency,
    now: Date = Date(),
    calendar: Calendar = .current
  ) -> String {
    let due = nextOccurrence(anchorDate: anchorDate, frequency: frequency, now: now, calendar: calendar)
    let today = calendar.startOfDay(for: now)
    let days = Int(round(due.timeIntervalSince(today) / 86_400))
    let formatter = DateFormatter()
    formatter.locale = .current
    formatter.setLocalizedDateFormatFromTemplate("MMMd")
    let date = formatter.string(from: due)
    let relative: String
    if days == 0 {
      relative = "today"
    } else if days == 1 {
      relative = "tomorrow"
    } else {
      relative = "in \(days) days"
    }
    return "Due \(date) · \(relative)"
  }
}

File: android-native/app/src/main/java/app/dimo/android/domain/DateHelpers.kt
package app.dimo.android.domain

import app.dimo.android.data.model.RecurringFrequency
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.time.temporal.ChronoUnit
import java.util.Locale

enum class RecurringOccurrenceSelection { ALL, SELECTED }

/**
 * Port of `ios-native/Dimo/Domain/DateHelpers.swift`.
 *
 * All month arithmetic is calendar-based (never fixed-millisecond intervals) so
 * short months and leap years land on the same dates as the Swift and TypeScript
 * implementations.
 */
object DateHelpers {
  const val DAY_MS = 86_400_000L

  fun zone(): ZoneId = ZoneId.systemDefault()

  fun localDate(timestamp: Long, zone: ZoneId = zone()): LocalDate =
    Instant.ofEpochMilli(timestamp).atZone(zone).toLocalDate()

  fun localDateKey(date: LocalDate): String =
    String.format(Locale.ROOT, "%04d-%02d-%02d", date.year, date.monthValue, date.dayOfMonth)

  fun localDateKey(timestamp: Long, zone: ZoneId = zone()): String =
    localDateKey(localDate(timestamp, zone))

  fun parseLocalDate(value: String): LocalDate {
    val parts = value.split("-").mapNotNull { it.toIntOrNull() }
    if (parts.size != 3) return LocalDate.now(zone())
    return runCatching { LocalDate.of(parts[0], parts[1], parts[2]) }
      .getOrElse { LocalDate.now(zone()) }
  }

  /** Midnight-of-day epoch millis, the analogue of `calendar.startOfDay`. */
  fun startOfDayMillis(date: LocalDate, zone: ZoneId = zone()): Long =
    date.atStartOfDay(zone).toInstant().toEpochMilli()

  fun formatTransactionDay(
    timestamp: Long,
    now: LocalDate = LocalDate.now(zone()),
    zone: ZoneId = zone(),
    locale: Locale = Locale.getDefault(),
  ): String {
    val date = localDate(timestamp, zone)
    if (date == now) return "Today"
    if (date == now.minusDays(1)) return "Yesterday"
    val pattern = if (date.year == now.year) "EEEE, MMM d" else "EEEE, MMM d, yyyy"
    return date.format(DateTimeFormatter.ofPattern(pattern, locale))
  }

  fun formatTransactionTime(
    timestamp: Long,
    zone: ZoneId = zone(),
    locale: Locale = Locale.getDefault(),
  ): String {
    val time = Instant.ofEpochMilli(timestamp).atZone(zone).toLocalTime()
    return time.format(DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT).withLocale(locale))
  }

  fun daysInMonth(year: Int, monthIndex: Int): Int {
    // monthIndex is zero-based, matching the Swift signature.
    val normalizedYear = year + Math.floorDiv(monthIndex, 12)
    val normalizedMonth = Math.floorMod(monthIndex, 12) + 1
    return LocalDate.of(normalizedYear, normalizedMonth, 1).lengthOfMonth()
  }

  /** The next occurrence on or after today, clamping the anchor day into short months. */
  fun nextOccurrence(
    anchorDate: String,
    frequency: RecurringFrequency,
    now: LocalDate = LocalDate.now(zone()),
  ): LocalDate {
    val anchor = parseLocalDate(anchorDate)
    if (!anchor.isBefore(now)) return anchor

    val anchorDay = anchor.dayOfMonth
    val anchorMonthIndex = anchor.monthValue - 1

    if (frequency == RecurringFrequency.MONTHLY) {
      fun candidate(year: Int, monthIndex: Int): LocalDate {
        val day = minOf(anchorDay, daysInMonth(year, monthIndex))
        return LocalDate.of(year, monthIndex + 1, day)
      }
      val year = now.year
      val monthIndex = now.monthValue - 1
      var result = candidate(year, monthIndex)
      if (result.isBefore(now)) {
        var nextMonth = monthIndex + 1
        var nextYear = year
        if (nextMonth > 11) {
          nextMonth = 0
          nextYear += 1
        }
        result = candidate(nextYear, nextMonth)
      }
      return result
    }

    fun candidate(year: Int): LocalDate {
      val day = minOf(anchorDay, daysInMonth(year, anchorMonthIndex))
      return LocalDate.of(year, anchorMonthIndex + 1, day)
    }
    var result = candidate(now.year)
    if (result.isBefore(now)) result = candidate(now.year + 1)
    return result
  }

  /** Every occurrence from the anchor up to and including today. */
  fun occurrencesThrough(
    anchorDate: String,
    frequency: RecurringFrequency,
    now: LocalDate = LocalDate.now(zone()),
  ): List<LocalDate> {
    val anchor = parseLocalDate(anchorDate)
    if (anchor.isAfter(now)) return emptyList()

    val day = anchor.dayOfMonth
    val dates = mutableListOf<LocalDate>()

    if (frequency == RecurringFrequency.MONTHLY) {
      var year = anchor.year
      var monthIndex = anchor.monthValue - 1
      while (dates.size < 1200) {
        val clamped = minOf(day, daysInMonth(year, monthIndex))
        val date = LocalDate.of(year, monthIndex + 1, clamped)
        if (date.isAfter(now)) break
        dates.add(date)
        monthIndex += 1
        if (monthIndex > 11) {
          monthIndex = 0
          year += 1
        }
      }
      return dates
    }

    var year = anchor.year
    val monthIndex = anchor.monthValue - 1
    while (dates.size < 200) {
      val clamped = minOf(day, daysInMonth(year, monthIndex))
      val date = LocalDate.of(year, monthIndex + 1, clamped)
      if (date.isAfter(now)) break
      dates.add(date)
      year += 1
    }
    return dates
  }

  fun recurringTransactionDates(
    anchorDate: String,
    frequency: RecurringFrequency,
    selection: RecurringOccurrenceSelection,
    now: LocalDate = LocalDate.now(zone()),
  ): List<LocalDate> {
    val anchor = parseLocalDate(anchorDate)
    if (anchor.isAfter(now)) return emptyList()
    return if (selection == RecurringOccurrenceSelection.ALL) {
      occurrencesThrough(anchorDate, frequency, now)
    } else {
      listOf(anchor)
    }
  }

  /** Occurrences pin to local noon so day grouping is stable across time zones. */
  fun occurrenceTimestamp(date: LocalDate, zone: ZoneId = zone()): Long =
    date.atTime(LocalTime.NOON).atZone(zone).toInstant().toEpochMilli()

  /** Occurrence on [date] carrying the hour and minute of [time]. */
  fun occurrenceTimestamp(date: LocalDate, time: LocalTime, zone: ZoneId = zone()): Long =
    LocalDateTime.of(date, LocalTime.of(time.hour, time.minute))
      .atZone(zone)
      .toInstant()
      .toEpochMilli()

  fun recurringDueLabel(
    anchorDate: String,
    frequency: RecurringFrequency,
    now: LocalDate = LocalDate.now(zone()),
    locale: Locale = Locale.getDefault(),
  ): String {
    val due = nextOccurrence(anchorDate, frequency, now)
    val days = ChronoUnit.DAYS.between(now, due).toInt()
    val date = due.format(DateTimeFormatter.ofPattern("MMM d", locale))
    val relative = when (days) {
      0 -> "today"
      1 -> "tomorrow"
      else -> "in $days days"
    }
    return "Due $date · $relative"
  }
}
